spatial/r3/mat.go

```diff
@@ -0,0 +1,268 @@
+func Eye() *Mat {
+	return &Mat{data: [3][3]float64{
+		{1, 0, 0},
+		{0, 1, 0},
+		{0, 0, 1},
+	}}
+}
+
+// Scale multiplies the elements of a by f, placing the result in the receiver.
+//
+// See the mat.Scaler interface for more information.
+func (m *Mat) Scale(f float64, a mat.Matrix) {
+	r, c := a.Dims()
+	if r != 3 || c != 3 {
+		panic(badDim)
+	}
+	for i := 0; i < 3; i++ {
+		for j := 0; j < 3; j++ {
+			m.Set(i, j, f*a.At(i, j))
+		}
+	}
+}
+
+// Performs matrix multiplication on v:
+//  result = M * v
+func (m *Mat) MulVec(v Vec) (result Vec) {
+	result.X = v.X*m.At(0, 0) + v.Y*m.At(0, 1) + v.Z*m.At(0, 2)
+	result.Y = v.X*m.At(1, 0) + v.Y*m.At(1, 1) + v.Z*m.At(1, 2)
+	result.Z = v.X*m.At(2, 0) + v.Y*m.At(2, 1) + v.Z*m.At(2, 2)
+	return result
+}
+
+// Performs transposed matrix multiplication on v:
+//  result = Mᵀ * v
+func (m *Mat) MulVecTrans(v Vec) (result Vec) {
+	result.X = v.X*m.At(0, 0) + v.Y*m.At(1, 0) + v.Z*m.At(2, 0)
+	result.Y = v.X*m.At(0, 1) + v.Y*m.At(1, 1) + v.Z*m.At(2, 1)
+	result.Z = v.X*m.At(0, 2) + v.Y*m.At(1, 2) + v.Z*m.At(2, 2)
+	return result
+}
+
+// Skew returns the 3×3 skew symmetric matrix (right hand system) of v.
+//                  [ 0 -z  y]
+//  Skew({x,y,z}) = [ z  0 -x]
+//                  [-y  x  0]
+func Skew(v Vec) (M *Mat) {
+	return NewMat([]float64{
+		0, -v.Z, v.Y,
+		v.Z, 0, -v.X,
+		-v.Y, v.X, 0,
+	})
+}
+
+// Mul takes the matrix product of a and b, placing the result in the receiver.
+// If the number of columns in a does not equal 3, Mul will panic.
+func (m *Mat) Mul(a, b mat.Matrix) {
+	ra, ca := a.Dims()
+	rb, cb := b.Dims()
+	switch {
+	case ra != 3:
+		panic(badDim)
+	case cb != 3:
+		panic(badDim)
+	case ca != rb:
+		panic(badDim)
+	}
+	if ca != 3 {
+		// General matrix multiplication for the case where the inner dimension is not 3.
+		t := mat.NewDense(3, 3, m.backingSlice())
+		t.Mul(a, b)
+		return
+	}
+	a00 := a.At(0, 0)
+	b00 := b.At(0, 0)
+	a01 := a.At(0, 1)
+	b01 := b.At(0, 1)
+	a02 := a.At(0, 2)
+	b02 := b.At(0, 2)
+	a10 := a.At(1, 0)
+	b10 := b.At(1, 0)
+	a11 := a.At(1, 1)
+	b11 := b.At(1, 1)
+	a12 := a.At(1, 2)
+	b12 := b.At(1, 2)
+	a20 := a.At(2, 0)
+	b20 := b.At(2, 0)
+	a21 := a.At(2, 1)
+	b21 := b.At(2, 1)
+	a22 := a.At(2, 2)
+	b22 := b.At(2, 2)
+	m.data[0][0] = a00*b00 + a01*b10 + a02*b20
+	m.data[0][1] = a00*b01 + a01*b11 + a02*b21
+	m.data[0][2] = a00*b02 + a01*b12 + a02*b22
+	m.data[1][0] = a10*b00 + a11*b10 + a12*b20
+	m.data[1][1] = a10*b01 + a11*b11 + a12*b21
+	m.data[1][2] = a10*b02 + a11*b12 + a12*b22
+	m.data[2][0] = a20*b00 + a21*b10 + a22*b20
+	m.data[2][1] = a20*b01 + a21*b11 + a22*b21
+	m.data[2][2] = a20*b02 + a21*b12 + a22*b22
+}
+
+// CloneFrom clones the matrix a into the receiver m. Mat expects a 3×3
+// input matrix.
+func (m *Mat) CloneFrom(a mat.Matrix) {
+	r, c := a.Dims()
+	if r != 3 || c != 3 {
+		panic(badDim)
+	}
+	m.cloneNested(a)
+}
+
+func (m *Mat) cloneNested(a mat.Matrix) {
+	for i := 0; i < 3; i++ {
+		for j := 0; j < 3; j++ {
```

Comment:
perhaps reverse the loops order? (for memory locality)

Reply:
This is iterating in the correct orientation for row major.

Reply:
d'oh. I was probably already a bit confused by the flu...